From 38b6917b8619ec896f49a01214c5f2b1b74e38f2 Mon Sep 17 00:00:00 2001
From: Yann Gautier <yann.gautier@foss.st.com>
Date: Thu, 20 Jun 2024 17:08:27 +0200
Subject: [PATCH 614/841] Revert "feat(stm32mp2): derivating DDRCryp encryption
 key from HUK"

This reverts commit dea534133cb40d570ed1b97ee88a3c4986db562c.
In case of TAMPER, we cannot use SAES and HUK derivation, so we won't
be able to boot in that case. Fall back to RNG use for the DDR
encryption key.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git b5e5f5d861aecc865c433a03433039623c0ed6f8 Revert "feat(stm32mp2): derivating DDRCryp encryption key from HUK"]
Change-Id: I76d1dec9f4a70ebc11ba441b866da97b3edb3948
Signed-off-by: Yann Gautier <yann.gautier@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/388907
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 plat/st/stm32mp2/bl2_plat_setup.c           | 43 +++------------------
 plat/st/stm32mp2/include/stm32mp2_context.h |  4 +-
 plat/st/stm32mp2/platform.mk                |  4 +-
 plat/st/stm32mp2/stm32mp2_context.c         | 12 +++---
 4 files changed, 16 insertions(+), 47 deletions(-)

diff --git a/plat/st/stm32mp2/bl2_plat_setup.c b/plat/st/stm32mp2/bl2_plat_setup.c
index ba5c1036b..8e3f44970 100644
--- a/plat/st/stm32mp2/bl2_plat_setup.c
+++ b/plat/st/stm32mp2/bl2_plat_setup.c
@@ -23,7 +23,6 @@
 #include <drivers/st/stm32_iwdg.h>
 #include <drivers/st/stm32_rifsc.h>
 #include <drivers/st/stm32_rng.h>
-#include <drivers/st/stm32_saes.h>
 #include <drivers/st/stm32_uart.h>
 #include <drivers/st/stm32mp_pmic2.h>
 #include <drivers/st/stm32mp_reset.h>
@@ -521,58 +520,28 @@ skip_console_init:
 	stm32mp_io_setup();
 }
 
-#if STM32MP_M33_TDCID
-static inline void prepare_encryption(void) {}
-#else /* !STM32MP_M33_TDCID */
-static int generate_enc_mkey_from_seed(uint8_t *mkey, uint8_t *seed, uint32_t seed_size){
-	int ret;
-	struct stm32_saes_context ctx;
-	uint8_t payload[RISAF_KEY_SIZE_IN_BYTES] = {0U};
-
-	assert(seed_size <= RISAF_KEY_SIZE_IN_BYTES);
-
-	ret = stm32_saes_driver_init();
-	if (ret != 0) {
-		return ret;
-	}
-
-	memcpy(payload, seed, seed_size); /* add seed in a block sized payload */
-
-	ret = stm32_saes_init(&ctx, false, STM32_SAES_MODE_ECB,
-			      STM32_SAES_KEY_DHU, NULL, RISAF_KEY_SIZE_IN_BYTES, NULL, 0U);
-	if (ret != 0) {
-		return ret;
-	}
-
-	return stm32_saes_update(&ctx, true, payload, mkey, sizeof(payload));
-}
-
 static void prepare_encryption(void)
 {
+#if !STM32MP_M33_TDCID
 	uint8_t mkey[RISAF_KEY_SIZE_IN_BYTES];
-	uint8_t seed[RISAF_SEED_SIZE_IN_BYTES];
 
 #if STM32MP_UART_PROGRAMMER || STM32MP_USB_PROGRAMMER
-	if (stm32_rng_read(seed, RISAF_SEED_SIZE_IN_BYTES) != 0) {
+	if (stm32_rng_read(mkey, RISAF_KEY_SIZE_IN_BYTES) != 0) {
 		panic();
 	}
 #else /* STM32MP_UART_PROGRAMMER || STM32MP_USB_PROGRAMMER */
 	if (stm32mp_is_wakeup_from_standby()) {
-		stm32mp_pm_get_enc_mkey_seed_from_context(seed);
+		stm32mp_pm_get_enc_mkey_from_context(mkey);
 	} else {
 		/* Generate RISAF encryption key from RNG */
-		if (stm32_rng_read(seed, RISAF_SEED_SIZE_IN_BYTES) != 0) {
+		if (stm32_rng_read(mkey, RISAF_KEY_SIZE_IN_BYTES) != 0) {
 			panic();
 		}
 
-		stm32mp_pm_save_enc_mkey_seed_in_context(seed);
+		stm32mp_pm_save_enc_mkey_in_context(mkey);
 	}
 #endif /* STM32MP_UART_PROGRAMMER || STM32MP_USB_PROGRAMMER */
 
-	if (generate_enc_mkey_from_seed(mkey, seed, RISAF_SEED_SIZE_IN_BYTES) != 0) {
-		panic();
-	}
-
 	if (stm32mp2_risaf_write_encryption_key(RISAF4_INST, mkey) != 0) {
 		panic();
 	}
@@ -582,8 +551,8 @@ static void prepare_encryption(void)
 		panic();
 	}
 #endif /* STM32MP21 */
+#endif /* !STM32MP_M33_TDCID */
 }
-#endif /* STM32MP_M33_TDCID */
 
 /*******************************************************************************
  * This function can be used by the platforms to update/use image
diff --git a/plat/st/stm32mp2/include/stm32mp2_context.h b/plat/st/stm32mp2/include/stm32mp2_context.h
index 46489ddc1..83ea4ac2d 100644
--- a/plat/st/stm32mp2/include/stm32mp2_context.h
+++ b/plat/st/stm32mp2/include/stm32mp2_context.h
@@ -11,8 +11,8 @@
 
 #include <lib/psci/psci.h>
 
-void stm32mp_pm_save_enc_mkey_seed_in_context(uint8_t *data);
-void stm32mp_pm_get_enc_mkey_seed_from_context(uint8_t *data);
+void stm32mp_pm_save_enc_mkey_in_context(uint8_t *data);
+void stm32mp_pm_get_enc_mkey_from_context(uint8_t *data);
 
 bool stm32_pm_context_is_valid(void);
 void stm32_pm_context_save(const psci_power_state_t *state);
diff --git a/plat/st/stm32mp2/platform.mk b/plat/st/stm32mp2/platform.mk
index 0c4f306b6..e7059255e 100644
--- a/plat/st/stm32mp2/platform.mk
+++ b/plat/st/stm32mp2/platform.mk
@@ -267,8 +267,7 @@ PLAT_BL_COMMON_SOURCES		+=	drivers/st/iwdg/stm32_iwdg.c				\
 
 PLAT_BL_COMMON_SOURCES		+=	drivers/st/clk/clk-stm32-core.c				\
 					drivers/st/clk/clk-stm32mp2.c				\
-					drivers/st/crypto/stm32_rng.c				\
-					drivers/st/crypto/stm32_saes.c
+					drivers/st/crypto/stm32_rng.c
 
 PLAT_BL_COMMON_SOURCES  	+=	drivers/st/nvmem/stm32mp_tamp_nvram_mp2.c
 
@@ -287,6 +286,7 @@ endif
 
 ifeq (${TRUSTED_BOARD_BOOT},1)
 BL2_SOURCES			+=	drivers/st/crypto/stm32_pka.c
+BL2_SOURCES			+=	drivers/st/crypto/stm32_saes.c
 endif
 
 ifneq ($(filter 1,${STM32MP_EMMC} ${STM32MP_SDMMC}),)
diff --git a/plat/st/stm32mp2/stm32mp2_context.c b/plat/st/stm32mp2/stm32mp2_context.c
index f040f37dc..4e288cc88 100644
--- a/plat/st/stm32mp2/stm32mp2_context.c
+++ b/plat/st/stm32mp2/stm32mp2_context.c
@@ -19,14 +19,14 @@
 
 #define BACKUP_CTX_ADDR		STM32MP_BACKUP_RAM_BASE
 #define BACKUP_CTX_CLK		CK_BUS_BKPSRAM
-#define KEY_SEED_SIZE_IN_BYTES	RISAF_SEED_SIZE_IN_BYTES
+#define ENC_KEY_SIZE_IN_BYTES	RISAF_KEY_SIZE_IN_BYTES
 
 /* Magic used to indicated valid = ' ' 'M' 'P' '2' */
 #define CONTEXT_MAGIC			0x204D5032
 
 struct backup_data_s {
 	uint32_t magic;
-	uint8_t mkey_seed[KEY_SEED_SIZE_IN_BYTES];
+	uint8_t enc_mkey[ENC_KEY_SIZE_IN_BYTES];
 	psci_power_state_t standby_pwr_state;
 	cpu_context_t saved_cpu_s_context[PLATFORM_CORE_COUNT];
 	cpu_context_t saved_cpu_ns_context[PLATFORM_CORE_COUNT];
@@ -38,7 +38,7 @@ struct backup_data_s {
 	uintptr_t fdt_bl31;
 };
 
-void stm32mp_pm_save_enc_mkey_seed_in_context(uint8_t *data)
+void stm32mp_pm_save_enc_mkey_in_context(uint8_t *data)
 {
 	struct backup_data_s *backup_data;
 
@@ -46,12 +46,12 @@ void stm32mp_pm_save_enc_mkey_seed_in_context(uint8_t *data)
 
 	clk_enable(BACKUP_CTX_CLK);
 
-	memcpy(backup_data->mkey_seed, data, KEY_SEED_SIZE_IN_BYTES);
+	memcpy(backup_data->enc_mkey, data, ENC_KEY_SIZE_IN_BYTES);
 
 	clk_disable(BACKUP_CTX_CLK);
 }
 
-void stm32mp_pm_get_enc_mkey_seed_from_context(uint8_t *data)
+void stm32mp_pm_get_enc_mkey_from_context(uint8_t *data)
 {
 	struct backup_data_s *backup_data;
 
@@ -59,7 +59,7 @@ void stm32mp_pm_get_enc_mkey_seed_from_context(uint8_t *data)
 
 	clk_enable(BACKUP_CTX_CLK);
 
-	memcpy(data, backup_data->mkey_seed, KEY_SEED_SIZE_IN_BYTES);
+	memcpy(data, backup_data->enc_mkey, ENC_KEY_SIZE_IN_BYTES);
 
 	clk_disable(BACKUP_CTX_CLK);
 }
-- 
2.47.2

